commit 5959f6c914bd5e9e9fbd858c8081ee19845df1b8
Author: Alexey Serbin <alexey@apache.org>
Date:   Mon Dec 9 22:03:37 2024 -0800

    ntp: adapt SO_REUSEPORT of NTP server socket for Kudu test scaffolding
    
    This patch makes the NTP server socket capable of binding to a port
    which is already bound, but not yet listened to.  This is useful
    in scenarios when it's necessary to make chronyd serving NTP requests
    at a port which has been reserved for chronyd by some other process.

diff --git a/socket.c b/socket.c
index 5b22db5..e09275b 100644
--- a/socket.c
+++ b/socket.c
@@ -456,10 +456,12 @@ bind_ip_address(int sock_fd, IPSockAddr *addr, int flags)
   if (addr->port > 0 && !SCK_SetIntOption(sock_fd, SOL_SOCKET, SO_REUSEADDR, 1))
     ;
 
-#if defined(LINUX) && defined(SO_REUSEPORT)
-  /* Allow multiple instances to bind to the same port in order to enable load
-     balancing.  Don't enable this option on non-Linux systems as it has
-     a slightly different meaning there (with some important implications). */
+#if defined(SO_REUSEPORT)
+  /* Make the socket capable of binding to a port which is already bound,
+   * but not yet listened to. This is useful in scenarios when it's necessary
+   * to make chronyd serving NTP requests at a port which has been reserved
+   * for chronyd by some other process.
+   */
   if (addr->port > 0 && !SCK_SetIntOption(sock_fd, SOL_SOCKET, SO_REUSEPORT, 1))
     ;
 #endif
@@ -470,10 +472,6 @@ bind_ip_address(int sock_fd, IPSockAddr *addr, int flags)
     ;
 #endif
 
-  /* Do not attempt to bind pre-initialised reusable socket */
-  if (SCK_IsReusable(sock_fd))
-    return 1;
-
   saddr_len = SCK_IPSockAddrToSockaddr(addr, (struct sockaddr *)&saddr, sizeof (saddr));
   if (saddr_len == 0)
     return 0;
